### PR TITLE
feat(php-version): Check php version in config

### PR DESCRIPTION
**Pull request by olivernybroe**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Fixed tickets | #346 

This PR adds support for checking composer file directly in the preset.
This enables the support for checking if we should enable `enableNativeTypeHint` when php version is lower than 7.4 in composer file.

The initial work for it is ready, however right now it only supports checking the php version if the version is formatted as `7.1` or `^7.1`, I haven't been able to find a way to make it work for all types yet without having to write the code for all scenarios.

### CODE REVIEW

File: src/Application/Composer.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public function lowestPhpVersionIsGreaterThenOrEqualTo(string $version): ?bool
+    {
+        $composerVersion = $this->getPhpVersion();
+
+        if ($composerVersion === null) {
+            return null;
+        }
+        $composerVersion = Str::after($composerVersion, '^');
```
Comment (Jibbarth) [inline src/Application/Composer.php]: Be careful. Illuminate/support is only in `require-dev`.

File: src/Application/DefaultPreset.php [hunk 2]
```diff
@@ -49,11 +50,16 @@ public static function get(): array
                 UnusedVariableSniff::class => [
                     'ignoreUnusedValuesWhenOnlyKeysAreUsedInForeach' => true,
                 ],
+                PropertyTypeHintSniff::class => [
+                    'enableNativeTypeHint' => optional($composer)->lowestPhpVersionMatches('7.4') === true ? false : null,
```
Comment (Jibbarth) [inline src/Application/DefaultPreset.php]: what is this function `optional` ?

File: src/Application/Composer.php (hunk 1 above)
Comment (olivernybroe) [reply src/Application/Composer.php]: Oh yeah, great catch, we can easily change this, just used to using it :)

File: src/Application/DefaultPreset.php (hunk 2 above)
Comment (olivernybroe) [reply src/Application/DefaultPreset.php]: It's again something from illuminate/support, it basically just says makes the method you are chaining on return null if the object you are wrapping in optional is null 👍 
Can easily be replaced with something not from illuminate/support

**Comment by olivernybroe** [comment]
@Jibbarth One thing is that I used `?Composer` a lot of places, because there might not be a composer file. However our phpstan settings disallows that. 
What do you propose here? I would think that either remove the rule or returning a Composer object with no data in it.

File: src/Application/Adapters/Drupal/Preset.php [hunk 3]
```diff
@@ -19,7 +20,7 @@ public static function getName(): string
         return 'drupal';
     }
 
-    public static function get(): array
+    public static function get(?Composer $composer): array
```
Comment (nunomaduro) [inline src/Application/Adapters/Drupal/Preset.php]: Is this syntax support in php 7.2 ?

Comment (olivernybroe) [reply src/Application/Adapters/Drupal/Preset.php]: It was part of 7.1 ([ref](https://www.php.net/releases/7_1_0.php)).

However our phpstan style does not like it, should we allow it or change it to not use that syntax?

**Comment by Jibbarth** [comment]
@olivernybroe :thinking:  I think I prefer a Composer object with empty data.

File: src/Application/Adapters/Drupal/Preset.php (hunk 3 above)
Comment (nunomaduro) [reply src/Application/Adapters/Drupal/Preset.php]: Can you explain me the context why we have `?Composer $composer`?

Comment (olivernybroe) [reply src/Application/Adapters/Drupal/Preset.php]: If we are not able to find a composer file, then Composer will be null. However I think I'll go ahead and change it to an Composer object with no data.

**Comment by olivernybroe** [comment]
@Jibbarth Just made a little change so it should be able to parse more version constraint strings.

File: src/Application/DefaultPreset.php [hunk 4]
```diff
@@ -49,11 +50,16 @@ public static function get(): array
                 UnusedVariableSniff::class => [
                     'ignoreUnusedValuesWhenOnlyKeysAreUsedInForeach' => true,
                 ],
+                PropertyTypeHintSniff::class => [
+                    'enableNativeTypeHint' => $composer->hasPhpVersion() ? $composer->lowestPhpVersionIsGreaterThenOrEqualTo('7.4') === true : null,
+                ],
             ],
         ];
+
+        return array_filter($config);
```
Comment (Jibbarth) [inline src/Application/DefaultPreset.php]: Is this `array_filter` really needed ? 

I think PropertyTypeHintSniff will be still here. 

Just do a quick test in php:

```
<?php

$arr = [
    'test' => 'true',
    'property' => ['test' => false]
];
    
var_dump($arr);
var_dump(array_filter($arr));
```

It's return the same array
```
array(2) { ["test"]=> string(4) "true" ["property"]=> array(1) { ["test"]=> bool(false) } }
array(2) { ["test"]=> string(4) "true" ["property"]=> array(1) { ["test"]=> bool(false) } }
```

Comment (olivernybroe) [reply src/Application/DefaultPreset.php]: Yeah, good point, was a mistake from my side :) Fixed it.

**Comment by olivernybroe** [comment]
@Jibbarth Might be possible yeah 😄 
But I agree that we should wait with that for now. 

Nice catch, was a `array_filter` I didn't need, should be good for a final review now.

**Pull request merged**